From: Tobias Heider <me@tobhe.de>
Date: Tue, 4 Apr 2023 16:59:05 +0200
Subject: Pass root_uuid directly to update-grub to prevent a race on
 first-boot where update-grub sometimes selects the old uuid.

---
 first-boot  | 4 +---
 update-grub | 7 ++++++-
 2 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/first-boot b/first-boot
index f114349..aaccad1 100755
--- a/first-boot
+++ b/first-boot
@@ -43,6 +43,4 @@ EOF
     echo
 fi
 
-update-grub
-
-cleanup
+update-grub "$root_uuid"
diff --git a/update-grub b/update-grub
index b94323d..1a83827 100755
--- a/update-grub
+++ b/update-grub
@@ -12,7 +12,12 @@ MODULES="ext2 part_gpt search"
 
 [ -e /etc/default/update-grub ] && source /etc/default/update-grub
 
-uuid="$(grub-probe "$BOOT_PART" -t fs_uuid)"
+# XXX: On first boot the UUID is passed via arg to avoid race condition
+if [ -z "$1" ]; then
+    uuid="$(grub-probe "$BOOT_PART" -t fs_uuid)"
+else
+    uuid="$1"
+fi
 part="$(grub-probe "$BOOT_PART" -t drive | sed -e 's/(.*,/hd0,/' | tr -d ')')"
 
 if [ -z "$uuid" ]; then
